webapp-django/crashstats/crashstats/views.py

```diff
@@ -817,24 +893,293 @@ def append_row_blob(blob, labels):
+
+    params.pop('date_end')
+
+    params.pop('date_start')
+    if _date_range_type == 'build':
+        params['build_id'] = [
+            '>=' + start_date.strftime('%Y%m%d000000'),
+            '<' + end_date.strftime('%Y%m%d000000'),
+        ]
+    else:
+        params['date'] = [
+            '>=' + start_date.strftime('%Y-%m-%d'),
+            '<' + end_date.strftime('%Y-%m-%d'),
+        ]
+
+    # XXX SuperSearch always returns `hits` even if you're only interested
+    # in facets.
+    # Setting this to 0 means it returns the default number of hits.
+    # Adrian, do we have a bug for this?
+    # Setting it to 1 just means it returns the minimal.
+    params['_results_number'] = 1  # because we don't care about hits
+
+    params['_columns'] = ('date', 'version', 'platform', 'product')
+    if params['hang_type'] == 'crash':
+        params['hang_type'] = '0'
+    elif params['hang_type'] == 'hang-p':
+        params['hang_type'] = '-1'
+    else:
+        params.pop('hang_type')
+
+    # supersearch expects the parameter `versions` (a list or tuple)
+    # to be called `version`
+    supersearch_params = copy.deepcopy(params)
+    supersearch_params['version'] = supersearch_params.pop('versions')
+    # in SuperSearch it's called 'Mac' not 'Mac OS X'
+    if 'Mac OS X' in supersearch_params['platforms']:
+        supersearch_params['platforms'].append('Mac')
+        supersearch_params['platforms'].remove('Mac OS X')
+    results = api.get(**supersearch_params)
+
+    # now get the ADI for these product versions
+    api = models.ADI()
+    adi_counts = api.get(
+        product=params['product'],
+        versions=params['versions'],
+        start_date=start_date,
+        end_date=end_date,
+        platforms=params['platforms'],
+    )
+
+    api = models.ProductBuildTypes()
+    product_build_types = api.get(product=params['product'])['build_types']
+
+    # This `adi_counts` is a list of dicts that looks like this:
+    #    {
+    #       'adi_count': 123,
+    #       'date': '2015-08-15',
+    #       'version': '40.0.2'
+    #       'build_type': 'beta',
+    #    }
+    # We need to turn this around so that it's like this:
+    #   {
+    #       '40.0.2': {
+    #           '2015-08-15': [123, 1.0],
+    #            ...
+    #       },
+    #       ...
+    #   }
+    # So it can easily be looked up how many counts there are per
+    # version per date.
+    #
+    # Note!! that the 1.0 in the example above is the throttle value
+    # for this build_type. We got that from the ProductBuildTypes API.
+    adi_by_version = {}
+    for group in adi_counts['hits']:
+        version = group['version']
+        build_type = group['build_type']
+        throttle = product_build_types[build_type]
+        if version not in adi_by_version:
+            adi_by_version[version] = {}
+        date = group['date']
+        count = group['adi_count']
+        adi_by_version[version][date] = [count, throttle]
+
+    cadu = {}
```

Comment:
I don't understand this variable name, I thus believe it is not a very good one. :->